### ISSUE TITLE
Dispatch `index_put` with boolean mask argument to masked_fill/masked_scatter when possible

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#57515]
Patterns like 
```
a[mask] = val
```
or
```
a[mask] = tensor
```
go through index_put, and convert boolean mask to Long indices. This is memory-inefficient and slow. Additionally, when people ask for deterministic algorithms, index_put_ does an unnecessary and inefficient thing of dispatching to sort-based algo (it's not needed for mask indexing), and sort-based algo has a bug related to broadcasting (see #61032).

Dispatch them to `masked_fill`/`masked_scatter` when possible (those functions don't convert mask to long indices). 
Should fix #53070 (although #53070 can be fixed directly). 
As the next step, `masked_select` should be efficiently implemented (instead of calling into `index`), and `index` should dispatch to `masked_select` when possible (should fix #30246)

For cpu and gpu, perf difference is significant, especially for multi-d tensors:
```
import torch
from torch.utils.benchmark import Timer, Compare

results = []
for device in ['cuda', 'cpu']:
    for dtype in [torch.float32]:
        for tensor_size in [100_000, 500000, 1_000_000, 5000000, 10_000_000]:
            if tensor_size >5000000 and device=='cpu':
                continue
            for mask_true_ratio in [0, 0.25, 0.5, 0.75, 1.]:
                a = torch.rand(tensor_size, dtype=dtype, device=device)
                mask = torch.empty(tensor_size, dtype=torch.bool, device=device).bernoulli_(mask_true_ratio)
                size = (1000, tensor_size//1000)
                val=1.
                stmts = ("a[mask]=val", "a.masked_fill_(mask, val)")
                labels = (f"index, {device}", f"masked_fill, {device}")
                timers = [Timer(stmt=stmt, num_threads = 10, sub_label=f"{tensor_size}, full {mask_true_ratio}", label = "Fill", description = l,
                globals = globals()) for stmt, l in zip(stmts, labels)]
                a=a.reshape(size)
                mask = mask.reshape(size)
                timers = timers + [Timer(stmt=stmt, num_threads = 10, sub_label=f"{size}, full {mask_true_ratio}", label = "Fill", description = l,
                globals = globals()) for stmt, l in zip(stmts, labels)]
                for t in timers:
                    results.append(t.blocked_autorange())

c=Compare(results)
c.print()
```
results on V100/devgpu, after #58468 fix. 
```
[------------------------------------------------ Fill ------------------------------------------------]
                                |  index, cuda  |  masked_fill, cuda  |  index, cpu  |  masked_fill, cpu
10 threads: --------------------------------------------------------------------------------------------
      100000, full 0            |      49.3     |          5.8        |      53.0    |         34.6     
      (1000, 100), full 0       |      49.7     |          5.4        |      61.6    |         34.5     
      100000, full 0.25         |      59.6     |          5.8        |     425.8    |        106.2     
      (1000, 100), full 0.25    |      62.4     |          5.9        |     486.8    |        100.2     
      100000, full 0.5          |      58.0     |          5.7        |     603.1    |        147.3     
      (1000, 100), full 0.5     |      61.3     |          5.8        |     713.8    |        132.0     
      100000, full 0.75         |      55.7     |          5.6        |     521.5    |        104.0     
      (1000, 100), full 0.75    |      61.4     |          5.7        |     656.8    |        103.5     
      100000, full 1.0          |      56.2     |          5.7        |     377.7    |         37.9     
      (1000, 100), full 1.0     |      60.3     |          5.7        |     502.5    |         33.7     
      500000, full 0            |      50.3     |          5.6        |     251.0    |         66.0     
      (1000, 500), full 0       |      52.7     |          9.2        |     264.0    |         60.9     
      500000, full 0.25         |      66.5     |          5.9        |    2137.1    |        217.6     
      (1000, 500), full 0.25    |      62.2     |          5.5        |    2342.0    |        210.0     
      500000, full 0.5          |      54.9     |          5.3        |    3065.1    |        312.0     
      (1000, 500), full 0.5     |      59.8     |          5.5        |    3445.2    |        310.9     
      500000, full 0.75         |      54.8     |          5.4        |    2551.4    |        225.6     
      (1000, 500), full 0.75    |      67.0     |          5.9        |    3292.6    |        209.1     
      500000, full 1.0          |      58.9     |          6.0        |    1814.8    |         61.1     
      (1000, 500), full 1.0     |      71.2     |          5.8        |    2475.1    |         60.7     
      1000000, full 0           |      52.5     |          7.4        |     492.8    |        128.8     
      (1000, 1000), full 0      |      52.3     |          7.4        |     486.3    |        121.1     
      1000000, full 0.25        |      57.4     |          7.4        |    4715.0    |        400.7     
      (1000, 1000), full 0.25   |      70.3     |          7.5        |    4806.8    |        389.3     
      1000000, full 0.5         |      65.2     |          7.4        |    6276.0    |        582.2     
      (1000, 1000), full 0.5    |      86.9     |          7.4        |    6968.4    |        628.1     
      1000000, full 0.75        |      66.9     |          7.5        |    5546.4    |        404.7     
      (1000, 1000), full 0.75   |     105.0     |          7.4        |    6085.3    |        408.8     
      1000000, full 1.0         |      65.7     |          7.4        |    3710.9    |        122.7     
      (1000, 1000), full 1.0    |     115.8     |          7.4        |    5175.9    |        120.4     
      5000000, full 0           |      79.4     |         56.6        |    2464.8    |        669.8     
      (1000, 5000), full 0      |      77.6     |         56.6        |    2412.8    |        542.6     
      5000000, full 0.25        |     162.2     |         56.5        |   20713.9    |       2042.4     
      (1000, 5000), full 0.25   |     216.5     |         56.5        |   22425.2    |       1929.4     
      5000000, full 0.5         |     183.0     |         56.3        |   30667.5    |       2964.3     
      (1000, 5000), full 0.5    |     296.5     |         56.3        |   53889.0    |       2935.9     
      5000000, full 0.75        |     202.2     |         56.9        |   25048.1    |       2046.9     
      (1000, 5000), full 0.75   |     359.2     |         56.8        |   62808.6    |       2003.0     
      5000000, full 1.0         |     191.8     |         56.8        |   37539.0    |        584.2     
      (1000, 5000), full 1.0    |     400.6     |         56.8        |   66521.3    |        549.5     
      10000000, full 0          |     143.2     |        114.0        |              |                  
      (1000, 10000), full 0     |     142.0     |        110.3        |              |                  
      10000000, full 0.25       |     285.2     |        110.3        |              |                  
      (1000, 10000), full 0.25  |     388.2     |        110.2        |              |                  
      10000000, full 0.5        |     322.0     |        109.9        |              |                  
      (1000, 10000), full 0.5   |     545.8     |        109.9        |              |                  
      10000000, full 0.75       |     362.2     |        110.2        |              |                  
      (1000, 10000), full 0.75  |     672.1     |        110.2        |              |                  
      10000000, full 1.0        |     363.0     |        109.9        |              |                  
      (1000, 10000), full 1.0   |     806.5     |        109.9        |              |                  

```
Benchmarks for index[mask]=values vs masked_scatter. Interestingly, on cpu, even though masked_scatter is a less general operation than indexing, it is always slower. On the gpu, masked_scatter is faster for denser masks (number of True elements >=.5 of all elements) and slower for emptier masks, due to how kernels are written (index scales more or less with the number of output elements, masked_scatter with the number of input). Unfortunately, without examining the mask (that's additional overhead), it's impossible to say which path should be taken
```
[-------------------------------------------------------- Fill --------------------------------------------------------]
                                |  index_vals, cuda  |  masked_scatter, cuda  |  index_vals, cpu  |  masked_scatter, cpu
10 threads: ------------------------------------------------------------------------------------------------------------
      100000, full 0            |        40.9        |          29.0          |         31.6      |           62.0      
      (1000, 100), full 0       |        46.1        |          28.7          |         34.0      |           65.3      
      100000, full 0.25         |        50.2        |          29.4          |        221.0      |          327.2      
      (1000, 100), full 0.25    |        53.6        |          28.8          |        237.7      |          323.6      
      100000, full 0.5          |        48.3        |          28.8          |        336.4      |          595.8      
      (1000, 100), full 0.5     |        53.5        |          28.4          |        347.9      |          520.3      
      100000, full 0.75         |        47.9        |          28.3          |        263.3      |          368.0      
      (1000, 100), full 0.75    |        53.1        |          28.9          |        295.6      |          371.3      
      100000, full 1.0          |        48.6        |          28.2          |        167.3      |          196.3      
      (1000, 100), full 1.0     |        52.2        |          28.4          |        210.0      |          201.0      
      500000, full 0            |        45.3        |          28.4          |         53.0      |          311.5      
      (1000, 500), full 0       |        43.8        |          28.0          |         56.4      |          326.8      
      500000, full 0.25         |        47.6        |          30.1          |        446.9      |         1844.7      
      (1000, 500), full 0.25    |        53.7        |          28.1          |        548.6      |         1663.1      
      500000, full 0.5          |        47.4        |          27.7          |        729.4      |         2632.6      
      (1000, 500), full 0.5     |        57.0        |          28.6          |        779.7      |         2631.8      
      500000, full 0.75         |        48.8        |          28.1          |        676.7      |         1841.5      
      (1000, 500), full 0.75    |        64.3        |          28.3          |        822.9      |         1847.9      
      500000, full 1.0          |        53.8        |          28.0          |        495.1      |          949.4      
      (1000, 500), full 1.0     |        73.7        |          28.0          |        585.1      |          952.0      
      1000000, full 0           |        41.5        |          46.4          |         95.4      |          618.8      
      (1000, 1000), full 0      |        43.2        |          46.3          |        101.3      |          648.1      
      1000000, full 0.25        |        57.9        |          48.2          |        904.8      |         3437.5      
      (1000, 1000), full 0.25   |        71.1        |          48.2          |       1008.7      |         3427.8      
      1000000, full 0.5         |        67.2        |          49.6          |       1389.5      |         5118.6      
      (1000, 1000), full 0.5    |        89.9        |          49.6          |       1705.0      |         5232.8      
      1000000, full 0.75        |        69.6        |          51.0          |       1270.0      |         3614.6      
      (1000, 1000), full 0.75   |       109.3        |          51.1          |       1505.6      |         3580.1      
      1000000, full 1.0         |        77.1        |          52.5          |        906.3      |         1921.0      
      (1000, 1000), full 1.0    |       123.5        |          52.5          |       1176.8      |         1937.6      
      5000000, full 0           |        76.4        |         179.1          |        457.8      |         3008.5      
      (1000, 5000), full 0      |        77.4        |         179.0          |        520.7      |         3374.6      
      5000000, full 0.25        |       166.3        |         185.8          |       5123.8      |        16038.4      
      (1000, 5000), full 0.25   |       224.6        |         185.8          |       4566.7      |        16114.6      
      5000000, full 0.5         |       194.3        |         191.7          |       6923.2      |        26119.1      
      (1000, 5000), full 0.5    |       308.3        |         191.7          |      12438.6      |        24990.6      
      5000000, full 0.75        |       220.9        |         198.6          |       6320.9      |        18308.8      
      (1000, 5000), full 0.75   |       393.7        |         198.6          |      14352.7      |        17785.5      
      5000000, full 1.0         |       218.9        |         204.8          |       9182.3      |         9708.5      
      (1000, 5000), full 1.0    |       429.2        |         204.7          |      15779.2      |        10091.9      
      10000000, full 0          |       140.8        |         342.7          |                   |                     
      (1000, 10000), full 0     |       137.3        |         342.7          |                   |                     
      10000000, full 0.25       |       294.6        |         356.7          |                   |                     
      (1000, 10000), full 0.25  |       408.6        |         356.7          |                   |                     
      10000000, full 0.5        |       349.5        |         369.2          |                   |                     
      (1000, 10000), full 0.5   |       575.3        |         369.2          |                   |                     
      10000000, full 0.75       |       402.6        |         382.3          |                   |                     
      (1000, 10000), full 0.75  |       716.5        |         403.1          |                   |                     
      10000000, full 1.0        |       407.9        |         394.6          |                   |                     
      (1000, 10000), full 1.0   |       859.7        |         394.5          |                   |                     

Times are in microseconds (us).
```
cc @VitalyFedyunin @ngimel

**Comment by ngimel** [comment]
Related #56896

**Comment by lgeiger** [comment]
@ngimel @eqy  I've come across this issue since I'm seeing bottlenecks due to an implicit CPU/GPU sync in `index_put` with boolean masks.

I re-ran the benchmarks comparing indexing with `masked_scatter` with a L40s GPU:
```python
import torch
from torch.utils.benchmark import Timer, Compare

results = []
for device in ['cuda', 'cpu']:
    for dtype in [torch.float32]:
        for tensor_size in [100_000, 500000, 1_000_000, 5000000, 10_000_000]:
            if tensor_size >5000000 and device=='cpu':
                continue
            for mask_true_ratio in [0, 0.25, 0.5, 0.75, 1.]:
                a = torch.rand(tensor_size, dtype=dtype, device=device)
                mask = torch.empty(tensor_size, dtype=torch.bool, device=device).bernoulli_(mask_true_ratio)
                size = (1000, tensor_size//1000)
                val = torch.rand(mask.sum().item(), dtype=dtype, device=device)
                stmts = ("a[mask]=val", "a.masked_scatter_(mask, val)")
                labels = (f"index, {device}", f"masked_scatter, {device}")
                timers = [Timer(stmt=stmt, num_threads = 10, sub_label=f"{tensor_size}, full {mask_true_ratio}", label = "Scatter", description = l,
                globals = globals()) for stmt, l in zip(stmts, labels)]
                a=a.reshape(size)
                mask = mask.reshape(size)
                timers = timers + [Timer(stmt=stmt, num_threads = 10, sub_label=f"{size}, full {mask_true_ratio}", label = "Scatter", description = l,
                globals = globals()) for stmt, l in zip(stmts, labels)]
                for t in timers:
                    results.append(t.blocked_autorange())

c=Compare(results)
c.print()
```

```
[------------------------------------------------- Scatter --------------------------------------------------]
                                |  index, cuda  |  masked_scatter, cuda  |  index, cpu  |  masked_scatter, cpu
10 threads: --------------------------------------------------------------------------------------------------
      100000, full 0            |      33.4     |          13.9          |     101.9    |           56.9
      (1000, 100), full 0       |      35.3     |          14.1          |     104.2    |           57.1
      100000, full 0.25         |      38.6     |          13.9          |     439.5    |          216.2
      (1000, 100), full 0.25    |      43.4     |          14.2          |     473.6    |          217.3
      100000, full 0.5          |      39.0     |          14.1          |     598.0    |          365.8
      (1000, 100), full 0.5     |      44.0     |          14.3          |     649.5    |          366.4
      100000, full 0.75         |      38.8     |          14.1          |     553.6    |          255.6
      (1000, 100), full 0.75    |      44.2     |          14.3          |     652.2    |          256.2
      100000, full 1.0          |      38.9     |          14.2          |     533.0    |          122.5
      (1000, 100), full 1.0     |      44.0     |          14.3          |     639.2    |          122.0
      500000, full 0            |      34.5     |          14.1          |     172.1    |          283.8
      (1000, 500), full 0       |      36.6     |          14.3          |     167.6    |          285.6
      500000, full 0.25         |      39.0     |          14.1          |    1188.7    |         1079.7
      (1000, 500), full 0.25    |      43.7     |          14.4          |    1307.2    |         1080.0
      500000, full 0.5          |      39.3     |          14.1          |    1801.2    |         1922.6
      (1000, 500), full 0.5     |      44.2     |          14.4          |    2056.8    |         1923.3
      500000, full 0.75         |      39.0     |          14.1          |    1751.5    |         1282.6
      (1000, 500), full 0.75    |      44.1     |          14.4          |    2139.4    |         1281.0
      500000, full 1.0          |      39.3     |          14.1          |    1635.8    |          605.5
      (1000, 500), full 1.0     |      44.4     |          14.4          |    2161.4    |          606.7
      1000000, full 0           |      34.7     |          15.7          |     242.1    |          565.4
      (1000, 1000), full 0      |      36.6     |          15.7          |     245.6    |          567.1
      1000000, full 0.25        |      39.0     |          15.9          |    2147.1    |         2155.6
      (1000, 1000), full 0.25   |      44.6     |          15.9          |    2407.8    |         2156.3
      1000000, full 0.5         |      39.2     |          16.2          |    3423.0    |         3838.3
      (1000, 1000), full 0.5    |      44.2     |          16.1          |    3958.2    |         3843.7
      1000000, full 0.75        |      39.2     |          16.3          |    3293.3    |         2563.5
      (1000, 1000), full 0.75   |      44.3     |          16.4          |    4108.5    |         2563.0
      1000000, full 1.0         |      39.0     |          16.4          |    3093.9    |         1210.3
      (1000, 1000), full 1.0    |      44.0     |          16.4          |    4106.8    |         1210.5
      5000000, full 0           |      42.5     |          40.7          |     929.8    |         2825.4
      (1000, 5000), full 0      |      45.6     |          40.7          |     928.5    |         2823.2
      5000000, full 0.25        |      56.6     |          43.5          |   10073.1    |        10781.0
      (1000, 5000), full 0.25   |      70.2     |          43.6          |   11478.4    |        10787.6
      5000000, full 0.5         |      58.6     |          44.5          |   15872.1    |        19242.0
      (1000, 5000), full 0.5    |      79.1     |          45.2          |   25293.3    |        19239.3
      5000000, full 0.75        |      64.9     |          45.7          |   15634.4    |        12811.6
      (1000, 5000), full 0.75   |     136.4     |          45.9          |   29114.5    |        12829.1
      5000000, full 1.0         |      65.1     |          46.3          |   21705.9    |         6118.6
      (1000, 5000), full 1.0    |     224.4     |          46.4          |   33864.0    |         6122.7
      10000000, full 0          |      61.3     |         269.8          |              |
      (1000, 10000), full 0     |      62.8     |         269.6          |              |
      10000000, full 0.25       |      84.3     |         295.2          |              |
      (1000, 10000), full 0.25  |     160.0     |         295.3          |              |
      10000000, full 0.5        |     205.2     |         319.8          |              |
      (1000, 10000), full 0.5   |     427.7     |         319.7          |              |
      10000000, full 0.75       |     350.6     |         341.5          |              |
      (1000, 10000), full 0.75  |     730.1     |         341.6          |              |
      10000000, full 1.0        |     302.6     |         356.4          |              |
      (1000, 10000), full 1.0   |     829.4     |         356.5          |              |

Times are in microseconds (us).
```

>  On the gpu, masked_scatter is faster for denser masks (number of True elements >=.5 of all elements) and slower for emptier masks, due to how kernels are written (index scales more or less with the number of output elements, masked_scatter with the number of input). Unfortunately, without examining the mask (that's additional overhead), it's impossible to say which path should be taken

I think these conclusions might not be valid anymore. Especially for smaller tensors it seems like `masked_scatter` is always faster. I assume this is because of the overhead of the implicit CPU-GPU sync. Only for very large tensors `masked_scatter` is sometimes slower for emptier masks.

How about always dispatching to `masked_scatter` unless the tensor is large? This would have the added benefit that this removes an implicit CPU-GPU sync which can be unexpected #12461

**Comment by lgeiger** [comment]
When running with `torch.use_deterministic_algorithms(True)` the difference is even larger:
```
[------------------------------------------------- Scatter --------------------------------------------------]
                                |  index, cuda  |  masked_scatter, cuda  |  index, cpu  |  masked_scatter, cpu
10 threads: --------------------------------------------------------------------------------------------------
      100000, full 0            |       40.4    |          17.0          |     101.8    |           56.9
      (1000, 100), full 0       |       46.5    |          17.2          |     103.7    |           57.0
      100000, full 0.25         |      194.4    |          17.2          |     429.2    |          215.9
      (1000, 100), full 0.25    |      227.1    |          17.4          |     576.0    |          216.2
      100000, full 0.5          |      178.2    |          17.3          |     698.5    |          366.6
      (1000, 100), full 0.5     |      235.5    |          17.6          |     854.6    |          366.3
      100000, full 0.75         |      181.5    |          17.4          |     721.0    |          256.1
      (1000, 100), full 0.75    |      236.6    |          17.6          |     932.8    |          255.5
      100000, full 1.0          |      186.1    |          17.3          |     743.7    |          121.6
      (1000, 100), full 1.0     |      249.3    |          17.6          |    1025.5    |          121.9
      500000, full 0            |       39.7    |          17.1          |     166.9    |          284.1
      (1000, 500), full 0       |       46.3    |          17.3          |     169.0    |          283.2
      500000, full 0.25         |      191.4    |          17.1          |    1458.3    |         1076.4
      (1000, 500), full 0.25    |      245.1    |          17.4          |    1822.3    |         1076.5
      500000, full 0.5          |      217.9    |          17.1          |    2340.8    |         1911.6
      (1000, 500), full 0.5     |      286.2    |          17.3          |    3064.7    |         1912.8
      500000, full 0.75         |      242.9    |          17.1          |    2526.7    |         1282.3
      (1000, 500), full 0.75    |      306.8    |          17.3          |    3583.1    |         1278.8
      500000, full 1.0          |      262.7    |          17.2          |    2640.1    |          600.3
      (1000, 500), full 1.0     |      329.6    |          17.3          |    4067.4    |          600.3
      1000000, full 0           |       40.4    |          20.0          |     249.2    |          566.6
      (1000, 1000), full 0      |       47.2    |          20.0          |     245.2    |          566.6
      1000000, full 0.25        |      217.4    |          20.3          |    2736.3    |         2149.3
      (1000, 1000), full 0.25   |      285.8    |          20.3          |    3424.0    |         2151.1
      1000000, full 0.5         |      260.0    |          20.2          |    4415.7    |         3842.9
      (1000, 1000), full 0.5    |      331.7    |          20.2          |    5973.2    |         3843.5
      1000000, full 0.75        |      315.1    |          20.4          |    4822.1    |         2556.9
      (1000, 1000), full 0.75   |      377.0    |          20.4          |    6897.7    |         2560.9
      1000000, full 1.0         |      364.5    |          20.3          |    5066.1    |         1199.9
      (1000, 1000), full 1.0    |      432.0    |          20.3          |    7895.5    |         1199.7
      5000000, full 0           |       42.3    |          55.9          |     911.0    |         2826.7
      (1000, 5000), full 0      |       48.9    |          55.9          |     919.2    |         2828.2
      5000000, full 0.25        |      418.6    |          57.4          |   12551.9    |        10777.6
      (1000, 5000), full 0.25   |      494.6    |          57.4          |   15992.4    |        10775.1
      5000000, full 0.5         |      805.6    |          58.0          |   20324.0    |        19230.6
      (1000, 5000), full 0.5    |      940.9    |          58.1          |   34679.7    |        19233.5
      5000000, full 0.75        |     1489.2    |          58.6          |   22364.2    |        12821.6
      (1000, 5000), full 0.75   |     1749.6    |          58.6          |   43350.8    |        12822.8
      5000000, full 1.0         |     2152.3    |          59.4          |   31121.4    |         6054.8
      (1000, 5000), full 1.0    |     2589.3    |          59.4          |   51319.7    |         6048.7
      10000000, full 0          |       65.6    |         363.6          |              |
      (1000, 10000), full 0     |       65.5    |         363.5          |              |
      10000000, full 0.25       |      883.3    |         402.9          |              |
      (1000, 10000), full 0.25  |     1002.7    |         403.0          |              |
      10000000, full 0.5        |     2250.6    |         421.3          |              |
      (1000, 10000), full 0.5   |     2664.1    |         421.4          |              |
      10000000, full 0.75       |     3591.0    |         448.8          |              |
      (1000, 10000), full 0.75  |     4530.0    |         448.8          |              |
      10000000, full 1.0        |     5000.3    |         467.6          |              |
      (1000, 10000), full 1.0   |     6541.1    |         467.7          |              |

Times are in microseconds (us).
```

**Comment by ngimel** [comment]
We'll accept a PR making this change

**Comment by ngimel** [comment]
On H100 on an abbreviated version of this benchmark I still see index faster for emptier masks:
```
[------------------------------ Scatter ------------------------------]
                                |  index, cuda  |  masked_scatter, cuda
10 threads: -----------------------------------------------------------
      5000000, full 0           |      34.5     |          63.2
      (1000, 5000), full 0      |      33.9     |          63.3
      5000000, full 0.25        |      50.2     |          66.6
      (1000, 5000), full 0.25   |      79.8     |          66.6
      5000000, full 0.5         |      72.9     |          69.4
      (1000, 5000), full 0.5    |     122.0     |          69.3
      10000000, full 0          |      46.5     |         142.4
      (1000, 10000), full 0     |      44.4     |         142.5
      10000000, full 0.25       |     107.2     |         147.6
      (1000, 10000), full 0.25  |     149.5     |         147.6
      10000000, full 0.5        |     136.6     |         153.0
      (1000, 10000), full 0.5   |     216.9     |         153.2
```

**Comment by lgeiger** [comment]
> On H100 on an abbreviated version of this benchmark I still see index faster for emptier masks:

Interesting, do you know why this is? Could it be that the H100 has faster host-GPU communication such that the host-GPU sync doesn't add as much overhead?

**Comment by ngimel** [comment]
`index_put` with a boolean mask is a pretty rare operation because you somehow have to know the number of true elements to have the correct rhs? It it's `tgt[mask]=src[mask]` you are better off doing `torch.where(mask, src, tgt)`

**Comment by lgeiger** [comment]
> `index_put` with a boolean mask is a pretty rare operation because you somehow have to know the number of true elements to have the correct rhs? It it's `tgt[mask]=src[mask]` you are better off doing `torch.where(mask, src, tgt)`

I agree, it's not a super common operation. In my case it's a bit more complicated as `torch.where` is not an option. I'm seeing the pattern in vllm doing something like `tgt[mask]=network(src[mask])` not sure whether this could be simplified.

But I just realised that in most of my cases mask is actually a CPU tensor for me so the GPU/host sync can be removed by converting the boolean mask to indices and then copying them to the GPU in a non-blocking way. I might send a PR to handle this automatically.
